node-wasm/src/worker.rs

```diff
@@ -0,0 +1,372 @@
+        match message {
+            WorkerMessage::NewConnection(connection) => {
+                message_server.add(connection);
+            }
+            WorkerMessage::Command((command, client_id)) => {
+                info!("received from {client_id:?}: {command:?}");
+                let Some(worker) = &mut worker else {
+                    match command {
+                        NodeCommand::IsRunning => {
+                            message_server.respond_to(client_id, WorkerResponse::IsRunning(false));
+                        }
+                        NodeCommand::StartNode(config) => {
+                            match NodeWorker::new(config).await {
+                                Ok(node) => {
+                                    worker = Some(node);
+                                    message_server
+                                        .respond_to(client_id, WorkerResponse::NodeStarted(Ok(())));
+                                }
+                                Err(e) => {
+                                    message_server
+                                        .respond_to(client_id, WorkerResponse::NodeStarted(Err(e)));
+                                }
+                            };
+                        }
+                        _ => {
+                            warn!("Worker not running");
+                            message_server.respond_err_to(client_id, WorkerError::NodeNotRunning);
+                        }
+                    }
+                    continue;
+                };
+
+                let response = worker.process_command(command).await;
+                message_server.respond_to(client_id, response);
+            }
+        }
+    }
+
+    error!("Channel to WorkerMessageServer closed, should not happen");
+}
+
+/// SharedWorker can only be spawned from an [`URL`]. To eliminate a need to host a js shim
+/// which calls into our Rust wasm code under specific path, we encode its entire contents
+/// as a [`Blob`] which is then passed as an URL.
+///
+/// [`URL`]: https://developer.mozilla.org/en-US/docs/Web/API/URL
+/// ['Blob']: https://developer.mozilla.org/en-US/docs/Web/API/Blob
+
+pub(crate) fn spawn_worker(name: &str, wasm_url: &str) -> Result<SharedWorker, JsError> {
+    let script = format!(
+        r#"
+Error.stackTraceLimit = 99;
+
+let queued = [];
+onconnect = (event) => {{
+  console.log("Queued connection", event);
+  queued.push(event.ports[0]);
+}}
+
+self.lumina = await import(self.location.origin + '{wasm_url}');
+await self.lumina.default();
+await self.lumina.run_worker(queued);
+"#
+    );
+
+    info!("js shim: {script}");
+
+    let array = Array::new();
+    array.push(&script.into());
+    let blob = Blob::new_with_str_sequence_and_options(
+        &array,
+        BlobPropertyBag::new().type_("application/javascript"),
+    )
+    .map_err(|e| JsError::new(&format!("could not create SharedWorker Blob: {e:?}")))?;
```

Comment:
we could replace those `.map_err` with `.js_context` unless we want to remove it completely

Reply:
Existing JsContext didn't exactly fit the need here, but as it's not used anyway I've modeled `JsValueToJsError` on it

Reply:
Why it doesn't fit the need? The error is not printed correctly with `Display`?

Reply:
Reworked Worker errors to use newly introduced js error types